dallinger/db.py

```diff
@@ -154,6 +155,91 @@ def init_db(drop_all=False, bind=engine):
     return session
 
 
+def get_all_mapped_classes():
+    """
+    Lists the different classes that are mapped with SQLAlchemy.
+    Classes are only included if they have at least one row in the database.
+    Returns a dictionary, keyed by class names,
+    where each value is itself a dictionary with three values:
+    ``cls``, the class itself;
+    ``table``, the database table within which the class can be found,
+    and ``polymorphic_identity``, the string label with which the class is
+    identified in the table's ``type`` column. The ``polymorphic_identity`` field
+    takes a value of ``None`` if the table does not use polymorphic inheritance.
+    """
+    classes = {}
+    for table in Base.metadata.tables.values():
+        if "type" in table.columns:
+            observed_types = [
+                r.type for r in session.query(table.columns.type).distinct().all()
+            ]
+            mapping = get_polymorphic_mapping(table)
+            for type_ in observed_types:
+                cls = mapping[type_]
+                classes[cls.__name__] = {
+                    "cls": cls,
+                    "table": table.name,
+                    "polymorphic_identity": type_,
+                }
+        else:
+            if session.query(table.columns.id).count() > 0:
+                cls = get_mapped_class(table)
+                classes[cls.__name__] = {
+                    "cls": cls,
+                    "table": table.name,
+                    "polymorphic_identity": None,
+                }
+    return classes
+
+
+def get_polymorphic_mapping(table: Union[str, Table]):
+    """
+    Gets the polymorphic mapping for a given table.
+    Returns a dictionary
+    where the dictionary keys correspond to polymorphic identities
+    (i.e. possible values of the table's ``type`` column)
+    and the dictionary values correspond to classes.
+    """
+    if isinstance(table, str):
+        table_name = table
+    else:
+        assert isinstance(table, Table)
+        table_name = table.name
+
+    return {
+        mapper.polymorphic_identity: mapper.class_
+        for mapper in Base.registry.mappers
+        if table_name in [table.name for table in mapper.tables]
+    }
+
+
+def get_mapped_classes(table: Union[str, Table]):
```

Comment:
There's quite some code duplication in `get_polymorphic_mapping` and `get_mapped_classes`

Reply:
Good point, I have now refactored out shared code.